managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java

```diff
@@ -1850,9 +1854,172 @@ public void offloadFailed(ManagedLedgerException e, Object ctx) {
+                } else {
+                    firstLedgerRetained = ls.getLedgerId();
+                    break;
+                }
+            }
+            firstUnoffloaded = PositionImpl.get(firstLedgerRetained, 0);
+        }
+
+        if (ledgersToOffload.isEmpty()) {
+            callback.offloadComplete(firstUnoffloaded, ctx);
+            return;
+        }
+
+        log.info("[{}] Going to offload ledgers {}", name,
+                 ledgersToOffload.stream().map(l -> l.getLedgerId()).collect(Collectors.toList()));
+        List< Pair<Long, CompletableFuture<byte[]>> > contexts = ledgersToOffload.stream()
+            .map(info -> {
+                    long ledgerId = info.getLedgerId();
+                    Map<String, String> extraMetadata = ImmutableMap.of("ManagedLedgerName", name);
+                    CompletableFuture<byte[]> context = getLedgerHandle(ledgerId).thenCompose(
+                            readHandle -> config.getLedgerOffloader().offload(readHandle, extraMetadata));
+                    return Pair.of(ledgerId, context);
+                }).collect(Collectors.toList());
+
+        CompletableFuture.allOf(contexts.stream().map(p -> p.getRight()).toArray(CompletableFuture[]::new))
+            .whenComplete((ignore, offloadException) -> {
+                    // If any of the offload operations failed offloadException will
+                    // be set. However some of the operations could have been successful.
+                    // If so, save the contexts for the successful prefix and notify the client
+                    // of the error if any occurred.
+                    List<Pair<Long, byte[]>> successfulOffloads = Lists.newArrayList();
+                    int errors = 0;
+                    synchronized (this) {
+                        ledgersListMutex.lock();
+
+                        // loop through results of offload operations. If an error occurred
+                        // check that the ledger still exists for the ML. If not, then it was
+                        // trimmed and the error can be ignored.
+                        for (Pair<Long, CompletableFuture<byte[]>> context : contexts) {
+                            if (context.getRight().isCompletedExceptionally()) {
+                                if (ledgers.containsKey(context.getLeft())) {
+                                    errors++;
+                                }
+                            } else {
+                                successfulOffloads.add(Pair.of(context.getLeft(), context.getRight().join()));
+                            }
+                        }
+                        ledgersListMutex.unlock();
+                    }
+                    log.info("[{}] All offload operations complete, {} successful, {} errors",
+                             name, successfulOffloads.size(), errors);
+
+                    int errorsToReport = errors; // effectively final so can be used in lambda
```

Comment:
nit: can't you just do 'final int errorsToReport = errors'? in this way it is self-commented.